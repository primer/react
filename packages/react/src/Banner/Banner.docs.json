{
  "id": "banner",
  "name": "Banner",
  "status": "alpha",
  "a11yReviewed": false,
  "importPath": "@primer/react/experimental",
  "stories": [],
  "props": [
    {
      "name": "aria-label",
      "type": "string",
      "description": "Provide an optional label to override the default name for the Banner landmark region"
    },
    {
      "name": "description",
      "type": "React.ReactNode",
      "description": "Provide an optional description for the Banner. This should provide supplemental information about the Banner"
    },
    {
      "name": "icon",
      "type": "React.ReactNode",
      "description": "Provide an icon for the banner"
    },
    {
      "name": "onDismiss",
      "type": "() => void",
      "description": "Optionally provide a handler to be called when the banner is dismissed. Providing this prop will show a dismiss button"
    },
    {
      "name": "primaryAction",
      "type": "React.ReactNode",
      "description": ""
    },
    {
      "name": "secondaryAction",
      "type": "React.ReactNode",
      "description": ""
    },
    {
      "name": "title",
      "type": "React.ReactNode",
      "description": "The title for the Banner. This will be used as the accessible name and is required unless `Banner.Title` is used as a child"
    },
    {
      "name": "variant",
      "type": "'critical' | 'info' | 'success' | 'upsell' | 'warning'",
      "description": ""
    }
  ],
  "subcomponents": [
    {
      "name": "Banner.Title",
      "props": [
        {
          "name": "as",
          "type": "'h2' | 'h3' | 'h4' | 'h5' | 'h6'"
        }
      ]
    },
    {
      "name": "Banner.Description",
      "props": []
    },
    {
      "name": "Banner.PrimaryAction",
      "props": []
    },
    {
      "name": "Banner.SecondaryAction",
      "props": []
    }
  ]
}
