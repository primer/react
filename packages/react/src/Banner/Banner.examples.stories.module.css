.RadioGroupWithTopMargin {
  margin-top: var(--base-size-16);
}
