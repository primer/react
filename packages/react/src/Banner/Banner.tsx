import cx from 'clsx'
import React, {useEffect} from 'react'
import styled from 'styled-components'
import {AlertIcon, InfoIcon, StopIcon, CheckCircleIcon, XIcon} from '@primer/octicons-react'
import {Button, IconButton} from '../Button'
import {get} from '../constants'
import {VisuallyHidden} from '../internal/components/VisuallyHidden'
import {useMergedRefs} from '../internal/hooks/useMergedRefs'

type BannerVariant = 'critical' | 'info' | 'success' | 'upsell' | 'warning'

export type BannerProps = React.ComponentPropsWithoutRef<'section'> & {
  /**
   * Provide an optional label to override the default name for the Banner
   * landmark region
   */
  'aria-label'?: string

  /**
   * Provide an optional description for the Banner. This should provide
   * supplemental information about the Banner
   */
  description?: React.ReactNode

  /**
   * Specify whether the title of the Banner should be visible or not.
   */
  hideTitle?: boolean

  /**
   * Provide an icon for the banner.
   * Note: Only `variant="info"` banners should use custom icons
   */
  icon?: React.ReactNode

  /**
   * Optionally provide a handler to be called when the banner is dismissed.
   * Providing this prop will show a dismiss button.
   *
   * Note: This is not available for critical banners.
   */
  onDismiss?: () => void

  /**
   * Provide an optional primary action for the Banner.
   */
  primaryAction?: React.ReactNode

  /**
   * Provide an optional secondary action for the Banner
   */
  secondaryAction?: React.ReactNode

  /**
   * The title for the Banner. This will be used as the accessible name and is
   * required unless `Banner.Title` is used as a child.
   */
  title?: React.ReactNode

  /**
   * Specify the type of the Banner
   */
  variant?: BannerVariant
}

const iconForVariant: Record<BannerVariant, React.ReactNode> = {
  critical: <StopIcon />,
  info: <InfoIcon />,
  success: <CheckCircleIcon />,
  upsell: <InfoIcon />,
  warning: <AlertIcon />,
}

const labels: Record<BannerVariant, string> = {
  critical: 'Critical',
  info: 'Information',
  success: 'Success',
  upsell: 'Recommendation',
  warning: 'Warning',
}

export const Banner = React.forwardRef<HTMLElement, BannerProps>(function Banner(
  {
    'aria-label': label,
    children,
    description,
    hideTitle,
    icon,
    onDismiss,
    primaryAction,
    secondaryAction,
    title,
    variant = 'info',
    ...rest
  },
  forwardRef,
) {
  const dismissible = variant !== 'critical' && onDismiss
  const hasActions = primaryAction || secondaryAction
  const bannerRef = React.useRef<HTMLElement>(null)
  const ref = useMergedRefs(forwardRef, bannerRef)

  if (__DEV__) {
    // This hook is called consistently depending on the environment
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useEffect(() => {
      if (title) {
        return
      }

      const {current: banner} = bannerRef
      if (!banner) {
        return
      }

      const hasTitle = banner.querySelector('[data-banner-title]')
      if (!hasTitle) {
        throw new Error(
          'Expected a title to be provided to the <Banner> component with the `title` prop or through `<Banner.Title>` but no title was found',
        )
      }
    }, [title])
  }

  return (
    <StyledBanner
      {...rest}
      aria-label={label ?? labels[variant]}
      as="section"
      data-dismissible={onDismiss ? '' : undefined}
      data-title-hidden={hideTitle ? '' : undefined}
      data-variant={variant}
      tabIndex={-1}
      ref={ref}
    >
      <style>{BannerContainerQuery}</style>
      <div className="BannerIcon">{icon && variant === 'info' ? icon : iconForVariant[variant]}</div>
      <div className="BannerContainer">
        <div className="BannerContent">
          {title ? (
            hideTitle ? (
              <VisuallyHidden>
                <BannerTitle>{title}</BannerTitle>
              </VisuallyHidden>
            ) : (
              <BannerTitle>{title}</BannerTitle>
            )
          ) : null}
          {description ? <BannerDescription>{description}</BannerDescription> : null}
          {children}
        </div>
        {hasActions ? <BannerActions primaryAction={primaryAction} secondaryAction={secondaryAction} /> : null}
      </div>
      {dismissible ? (
        <IconButton
          aria-label="Dismiss banner"
          onClick={onDismiss}
          className="BannerDismiss"
          icon={XIcon}
          variant="invisible"
        />
      ) : null}
    </StyledBanner>
  )
})

/**
 * For styling, it's important that the icons and the text have the same height
 * for alignment to occur in multi-line scenarios. Currently, we use a
 * line-height of `20px` so that means that the height of icons should match
 * that value.
 */
const StyledBanner = styled.div`
  display: grid;
  grid-template-columns: auto minmax(0, 1fr) auto;
  align-items: start;
  background-color: var(--banner-bgColor);
  border: var(--borderWidth-thin, 1px) solid var(--banner-borderColor);
  padding: var(--base-size-8, 0.5rem);
  border-radius: var(--borderRadius-medium, ${get('radii.2')});

  @supports (container-type: inline-size) {
    container: banner / inline-size;
  }

  &[data-variant='critical'] {
    --banner-bgColor: ${get('colors.danger.subtle')};
    --banner-borderColor: ${get('colors.danger.muted')};
    --banner-icon-fgColor: ${get('colors.danger.fg')};
  }

  &[data-variant='info'] {
    --banner-bgColor: ${get('colors.accent.subtle')};
    --banner-borderColor: ${get('colors.accent.muted')};
    --banner-icon-fgColor: ${get('colors.accent.fg')};
  }

  &[data-variant='success'] {
    --banner-bgColor: ${get('colors.success.subtle')};
    --banner-borderColor: ${get('colors.success.muted')};
    --banner-icon-fgColor: ${get('colors.success.fg')};
  }

  &[data-variant='upsell'] {
    --banner-bgColor: var(--bgColor-upsell-muted, ${get('colors.done.subtle')});
    --banner-borderColor: var(--borderColor-upsell-muted, ${get('colors.done.muted')});
    --banner-icon-fgColor: var(--fgColor-upsell-muted, ${get('colors.done.fg')});
  }

  &[data-variant='warning'] {
    --banner-bgColor: ${get('colors.attention.subtle')};
    --banner-borderColor: ${get('colors.attention.muted')};
    --banner-icon-fgColor: ${get('colors.attention.fg')};
  }

  /* BannerIcon ------------------------------------------------------------- */

  .BannerIcon {
    display: grid;
    place-items: center;
    padding: var(--base-size-8, 0.5rem);
  }

  .BannerIcon svg {
    color: var(--banner-icon-fgColor);
    fill: var(--banner-icon-fgColor);
    /* 20px is the line box height of the trailing action buttons */
    height: var(--base-size-20, 1.25rem);
  }

  &[data-title-hidden=''] .BannerIcon svg {
    height: var(--base-size-16, 1rem);
  }

  /* BannerContainer -------------------------------------------------------- */

  .BannerContainer {
    font-size: var(--text-body-size-medium, 0.875rem);
    align-items: start;
    line-height: var(--text-body-lineHeight-medium, calc(20 / 14));
    row-gap: var(--base-size-4, 0.25rem);
    column-gap: var(--base-size-4, 0.25rem);
  }

  & :where(.BannerContainer) {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
  }

  &[data-dismissible] .BannerContainer {
    display: grid;
    grid-template-columns: auto;
    grid-template-rows: auto;
  }

  /* BannerContent ---------------------------------------------------------- */

  .BannerContent {
    display: grid;
    row-gap: var(--base-size-4, 0.25rem);
    grid-column-start: 1;
    margin-block: var(--base-size-8, 0.5rem);
  }

  &[data-title-hidden=''] .BannerContent {
    margin-block: var(--base-size-6, 0.375rem);
  }

  @media screen and (min-width: 544px) {
    .BannerContent {
      flex: 1 1 0%;
    }
  }

  .BannerTitle {
    margin: 0;
    font-size: inherit;
    font-weight: var(--base-text-weight-semibold, 600);
  }

  /* BannerActions ---------------------------------------------------------- */
  .BannerActionsContainer {
    display: flex;
    column-gap: var(--base-size-8, 0.5rem);
    align-items: center;
  }

  .BannerActions :where([data-primary-action='trailing']) {
    display: none;
  }

  @media screen and (min-width: 544px) {
    .BannerActions :where([data-primary-action='trailing']) {
      display: flex;
    }

    .BannerActions :where([data-primary-action='leading']) {
      display: none;
    }
  }

  &[data-dismissible] .BannerActions {
    margin-block-end: var(--size-small, 0.375rem);
  }

  &[data-dismissible] .BannerActionsContainer[data-primary-action='trailing'] {
    display: none;
  }

  &[data-dismissible] .BannerActionsContainer[data-primary-action='leading'] {
    display: flex;
  }

  /* BannerDismiss ---------------------------------------------------------- */

  .BannerDismiss {
    display: grid;
    place-items: center;
    padding: var(--base-size-8, 0.5rem);
    margin-inline-start: var(--base-size-4, 0.25rem);
  }

  .BannerDismiss svg {
    color: var(--banner-icon-fgColor);
  }
`

const BannerContainerQuery = `
  @container banner (max-width: 500px) {
    .BannerContainer {
      display: grid;
      grid-template-rows: auto auto;
    }

    .BannerActions {
      margin-block-end: var(--size-small, 0.375rem);
    }

    .BannerActions [data-primary-action="trailing"] {
      display: none;
    }

    .BannerActions [data-primary-action="leading"] {
      display: flex;
    }
  }

  @container banner (min-width: 500px) {
    .BannerContainer {
      display: grid;
      grid-template-columns: auto auto;
    }

    .BannerActions [data-primary-action="trailing"] {
      display: flex;
    }

    .BannerActions [data-primary-action="leading"] {
      display: none;
    }
  }
`

type HeadingElement = 'h2' | 'h3' | 'h4' | 'h5' | 'h6'

export type BannerTitleProps<As extends HeadingElement> = {
  as?: As
  className?: string
} & React.ComponentPropsWithoutRef<As extends 'h2' ? 'h2' : As>

export function BannerTitle<As extends HeadingElement>(props: BannerTitleProps<As>) {
  const {as: Heading = 'h2', className, children, ...rest} = props
  return (
    <Heading {...rest} className={cx('BannerTitle', className)} data-banner-title="">
      {children}
    </Heading>
  )
}

export type BannerDescriptionProps = React.ComponentPropsWithoutRef<'div'>

export function BannerDescription({children, className, ...rest}: BannerDescriptionProps) {
  return (
    <div {...rest} className={cx('BannerDescription', className)}>
      {children}
    </div>
  )
}

export type BannerActionsProps = {
  primaryAction?: React.ReactNode
  secondaryAction?: React.ReactNode
}

export function BannerActions({primaryAction, secondaryAction}: BannerActionsProps) {
  return (
    <div className="BannerActions">
      <div className="BannerActionsContainer" data-primary-action="trailing">
        {secondaryAction ?? null}
        {primaryAction ?? null}
      </div>
      <div className="BannerActionsContainer" data-primary-action="leading">
        {primaryAction ?? null}
        {secondaryAction ?? null}
      </div>
    </div>
  )
}

export type BannerPrimaryActionProps = Omit<React.ComponentPropsWithoutRef<typeof Button>, 'variant'>

export function BannerPrimaryAction({children, className, ...rest}: BannerPrimaryActionProps) {
  return (
    <Button className={cx('BannerPrimaryAction', className)} variant="default" {...rest}>
      {children}
    </Button>
  )
}

export type BannerSecondaryActionProps = Omit<React.ComponentPropsWithoutRef<typeof Button>, 'variant'>

export function BannerSecondaryAction({children, className, ...rest}: BannerSecondaryActionProps) {
  return (
    <Button className={cx('BannerPrimaryAction', className)} variant="invisible" {...rest}>
      {children}
    </Button>
  )
}
