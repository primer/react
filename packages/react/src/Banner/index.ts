import {
  Banner as BannerImpl,
  BannerTitle,
  BannerDescription,
  BannerPrimaryAction,
  BannerSecondaryAction,
} from './Banner'
import type {
  BannerProps,
  BannerTitleProps,
  BannerDescriptionProps,
  BannerPrimaryActionProps,
  BannerSecondaryActionProps,
} from './Banner'

export const Banner = Object.assign(BannerImpl, {
  Title: BannerTitle,
  Description: BannerDescription,
  PrimaryAction: BannerPrimaryAction,
  SecondaryAction: BannerSecondaryAction,
})

export type {
  BannerProps,
  BannerTitleProps,
  BannerDescriptionProps,
  BannerPrimaryActionProps,
  BannerSecondaryActionProps,
}
