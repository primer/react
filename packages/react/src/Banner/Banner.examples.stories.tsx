import {Banner} from '../Banner'
import {action} from '@storybook/addon-actions'
import Link from '../Link'
import type {Meta} from '@storybook/react'
import {AriaAlert, AriaStatus} from '../live-region'
import FormControl from '../FormControl'
import RadioGroup from '../RadioGroup'
import Radio from '../Radio'
import {Button} from '../Button'
import React from 'react'
import {useFocus} from '../internal/hooks/useFocus'
import {PageLayout} from '../PageLayout'

const meta = {
  title: 'Drafts/Components/Banner/Examples',
  component: Banner,
} satisfies Meta<typeof Banner>

export default meta

export const WithUserAction = () => {
  const [hasError, setHasError] = React.useState(false)
  const bannerRef = React.useRef<React.ElementRef<typeof Banner>>(null)
  const focus = useFocus()

  return (
    <>
      {hasError ? (
        <Banner
          ref={bannerRef}
          title="Error"
          description={<AriaAlert>Something went wrong. Please try again later.</AriaAlert>}
          variant="critical"
        />
      ) : null}
      <Button
        type="button"
        onClick={() => {
          setHasError(true)
          focus(bannerRef)
        }}
      >
        Update profile
      </Button>
    </>
  )
}

export const WithDynamicContent = () => {
  type Choice = 'one' | 'two' | 'three'
  const messages: Map<Choice, string> = new Map([
    ['one', 'This is a message for choice one'],
    ['two', 'This is a message for choice two'],
    ['three', 'This is a message for choice three'],
  ])
  const [selected, setSelected] = React.useState<Choice>('one')

  return (
    <>
      <Banner
        title="Info"
        description={<AriaStatus>{messages.get(selected)}</AriaStatus>}
        onDismiss={action('onDismiss')}
        primaryAction={<Banner.PrimaryAction>Button</Banner.PrimaryAction>}
        secondaryAction={<Banner.SecondaryAction>Button</Banner.SecondaryAction>}
      />
      <RadioGroup
        sx={{marginTop: 4}}
        name="options"
        onChange={selected => {
          setSelected(selected as Choice)
        }}
      >
        <RadioGroup.Label>Choices</RadioGroup.Label>
        <FormControl>
          <Radio value="one" defaultChecked />
          <FormControl.Label>Choice one</FormControl.Label>
        </FormControl>
        <FormControl>
          <Radio value="two" />
          <FormControl.Label>Choice two</FormControl.Label>
        </FormControl>
        <FormControl>
          <Radio value="three" />
          <FormControl.Label>Choice three</FormControl.Label>
        </FormControl>
      </RadioGroup>
    </>
  )
}

export const WithCustomHeading = () => {
  return (
    <Banner
      onDismiss={action('onDismiss')}
      primaryAction={<Banner.PrimaryAction>Button</Banner.PrimaryAction>}
      secondaryAction={<Banner.SecondaryAction>Button</Banner.SecondaryAction>}
    >
      <Banner.Title as="h3">Info</Banner.Title>
      <Banner.Description>
        GitHub users are{' '}
        <Link inline underline href="#">
          now required
        </Link>{' '}
        to enable two-factor authentication as an additional security measure.
      </Banner.Description>
    </Banner>
  )
}

export const InSidebar = () => {
  return (
    <>
      <PageLayout>
        <PageLayout.Header divider="line">PageLayout Header</PageLayout.Header>
        <PageLayout.Pane position="start" divider="line">
          <h2>PageLayout Pane</h2>
          <Banner
            title="Info"
            description={
              <>
                GitHub users are{' '}
                <Link inline underline href="#">
                  now required
                </Link>{' '}
                to enable two-factor authentication as an additional security measure.
              </>
            }
            primaryAction={<Banner.PrimaryAction>Button</Banner.PrimaryAction>}
            secondaryAction={<Banner.SecondaryAction>Button</Banner.SecondaryAction>}
          />
        </PageLayout.Pane>
        <PageLayout.Content>
          <h1>PageLayout Content</h1>
        </PageLayout.Content>
      </PageLayout>
    </>
  )
}

export const Multiline = () => {
  return (
    <Banner
      onDismiss={action('onDismiss')}
      title="Info"
      description="GitHub users are now required to enable two-factor authentication as an additional security measure. Lorem IpsumÂ is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen bSed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora incidunt ut labore et dolore magnam aliquam quaerat voluptatem. Ut enim ad minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem vel eum iure reprehenderit qui in ea voluptate velit esse quam nihil molestiae consequatur, vel illum qui dolorem eum fugiat quo voluptas nulla pariatur?"
      primaryAction={<Banner.PrimaryAction>Button</Banner.PrimaryAction>}
      secondaryAction={<Banner.SecondaryAction>Button</Banner.SecondaryAction>}
    />
  )
}

export const DismissBanner = () => {
  const ref = React.useRef<React.ElementRef<'h2'>>(null)
  const [banner, setBanner] = React.useState<React.ComponentPropsWithoutRef<typeof Banner> | null>({
    title: 'Info',
    description: (
      <>
        GitHub users are{' '}
        <Link inline underline href="#">
          now required
        </Link>{' '}
        to enable two-factor authentication as an additional security measure.
      </>
    ),
  })

  return (
    <>
      <main>
        {banner ? (
          <Banner
            title={banner.title}
            description={banner.description}
            onDismiss={() => {
              setBanner(null)
              ref.current?.focus()
            }}
          />
        ) : null}
        <h2 ref={ref} tabIndex={-1}>
          Example page title
        </h2>
        <p>Example page content</p>
      </main>
    </>
  )
}
