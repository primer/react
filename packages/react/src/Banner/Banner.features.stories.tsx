import React from 'react'
import {CopilotIcon} from '@primer/octicons-react'
import {action} from '@storybook/addon-actions'
import type {Meta} from '@storybook/react'
import {Banner} from '../Banner'
import Link from '../Link'

const meta = {
  title: 'Drafts/Components/Banner/Features',
  component: Banner,
} satisfies Meta<typeof Banner>

export default meta

export const Critical = () => {
  return (
    <Banner
      title="Critical"
      description={
        <>
          GitHub users are{' '}
          <Link inline underline href="#">
            now required
          </Link>{' '}
          to enable two-factor authentication as an additional security measure.
        </>
      }
      variant="critical"
    />
  )
}

export const Info = () => {
  return (
    <Banner
      title="Info"
      description={
        <>
          GitHub users are{' '}
          <Link inline underline href="#">
            now required
          </Link>{' '}
          to enable two-factor authentication as an additional security measure.
        </>
      }
      onDismiss={action('onDismiss')}
      variant="info"
    ></Banner>
  )
}

export const Success = () => {
  return (
    <Banner
      title="Success"
      description={
        <>
          GitHub users are{' '}
          <Link inline underline href="#">
            now required
          </Link>{' '}
          to enable two-factor authentication as an additional security measure.
        </>
      }
      onDismiss={action('onDismiss')}
      variant="success"
    />
  )
}

export const Upsell = () => {
  return (
    <Banner
      title="Upsell"
      description={
        <>
          GitHub users are{' '}
          <Link inline underline href="#">
            now required
          </Link>{' '}
          to enable two-factor authentication as an additional security measure.
        </>
      }
      onDismiss={action('onDismiss')}
      variant="upsell"
    />
  )
}

export const Warning = () => {
  return (
    <Banner
      title="Warning"
      description={
        <>
          GitHub users are{' '}
          <Link inline underline href="#">
            now required
          </Link>{' '}
          to enable two-factor authentication as an additional security measure.
        </>
      }
      onDismiss={action('onDismiss')}
      variant="warning"
    />
  )
}

export const Dismiss = () => {
  return (
    <Banner
      title="Notice"
      description={
        <>
          GitHub users are{' '}
          <Link inline underline href="#">
            now required
          </Link>{' '}
          to enable two-factor authentication as an additional security measure.
        </>
      }
      onDismiss={action('onDismiss')}
    />
  )
}

export const DismissWithActions = () => {
  return (
    <Banner
      title="Notice"
      description={
        <>
          GitHub users are{' '}
          <Link inline underline href="#">
            now required
          </Link>{' '}
          to enable two-factor authentication as an additional security measure.
        </>
      }
      onDismiss={action('onDismiss')}
      primaryAction={<Banner.PrimaryAction>Button</Banner.PrimaryAction>}
      secondaryAction={<Banner.SecondaryAction>Button</Banner.SecondaryAction>}
    />
  )
}

export const WithHiddenTitle = () => {
  return (
    <Banner
      title="Warning"
      hideTitle
      description={
        <>
          GitHub users are{' '}
          <Link inline underline href="#">
            now required
          </Link>{' '}
          to enable two-factor authentication as an additional security measure.
        </>
      }
      variant="warning"
    />
  )
}

export const WithHiddenTitleAndActions = () => {
  return (
    <Banner
      title="Warning"
      hideTitle
      description={
        <>
          GitHub users are{' '}
          <Link inline underline href="#">
            now required
          </Link>{' '}
          to enable two-factor authentication as an additional security measure.
        </>
      }
      variant="warning"
      primaryAction={<Banner.PrimaryAction>Button</Banner.PrimaryAction>}
      secondaryAction={<Banner.SecondaryAction>Button</Banner.SecondaryAction>}
    />
  )
}

export const WithActions = () => {
  return (
    <Banner
      title="Warning"
      description={
        <>
          GitHub users are{' '}
          <Link inline underline href="#">
            now required
          </Link>{' '}
          to enable two-factor authentication as an additional security measure.
        </>
      }
      primaryAction={<Banner.PrimaryAction>Button</Banner.PrimaryAction>}
      secondaryAction={<Banner.SecondaryAction>Button</Banner.SecondaryAction>}
      variant="warning"
    />
  )
}

export const CustomIcon = () => {
  return (
    <Banner
      title="Upsell"
      description="An example banner with a custom icon"
      icon={<CopilotIcon />}
      onDismiss={action('onDismiss')}
      variant="upsell"
    />
  )
}
