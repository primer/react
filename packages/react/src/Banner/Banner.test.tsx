import {render, screen} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import {Banner} from '../Banner'

describe('Banner', () => {
  let spy: jest.SpyInstance

  beforeEach(() => {
    // Note: this error occurs due to our usage of `@container` within a
    // `<style>` tag in Banner. The CSS parser for jsdom does not support this
    // syntax and will fail with an error containing the message below.
    // eslint-disable-next-line no-console
    const originalConsoleError = console.error
    spy = jest.spyOn(console, 'error').mockImplementation((value, ...args) => {
      if (!value?.message?.includes('Could not parse CSS stylesheet')) {
        originalConsoleError(value, ...args)
      }
    })
  })

  afterEach(() => {
    spy.mockRestore()
  })

  it('should render as a region element', () => {
    render(<Banner title="test" />)
    expect(screen.getByRole('region', {name: 'Information'})).toBeInTheDocument()
    expect(screen.getByRole('heading', {name: 'test'})).toBeInTheDocument()
  })

  it('should label the landmark element with the corresponding variant label text', () => {
    render(<Banner title="test" />)
    expect(screen.getByRole('region')).toEqual(screen.getByLabelText('Information'))
  })

  it('should label the landmark element with the label for the critical variant', () => {
    render(<Banner title="test" variant="critical" />)
    expect(screen.getByRole('region')).toEqual(screen.getByLabelText('Critical'))
  })

  it('should label the landmark element with the label for the info variant', () => {
    render(<Banner title="test" variant="info" />)
    expect(screen.getByRole('region')).toEqual(screen.getByLabelText('Information'))
  })

  it('should label the landmark element with the label for the success variant', () => {
    render(<Banner title="test" variant="success" />)
    expect(screen.getByRole('region')).toEqual(screen.getByLabelText('Success'))
  })

  it('should label the landmark element with the label for the upsell variant', () => {
    render(<Banner title="test" variant="upsell" />)
    expect(screen.getByRole('region')).toEqual(screen.getByLabelText('Recommendation'))
  })

  it('should label the landmark element with the label for the warning variant', () => {
    render(<Banner title="test" variant="warning" />)
    expect(screen.getByRole('region')).toEqual(screen.getByLabelText('Warning'))
  })

  it('should support the `aria-label` prop to override the default label for the landmark', () => {
    render(<Banner aria-label="Test" title="test" variant="warning" />)
    expect(screen.getByRole('region')).toHaveAttribute('aria-label', 'Test')
  })

  it('should default the title to a h2', () => {
    render(<Banner title="test" />)
    expect(screen.getByRole('heading', {level: 2})).toBeInTheDocument()
    expect(screen.getByRole('heading', {level: 2})).toEqual(screen.getByText('test'))
  })

  it('should throw an error if no title is provided', () => {
    const spy = jest.spyOn(console, 'error').mockImplementation(() => {})
    expect(() => {
      render(<Banner />)
    }).toThrowErrorMatchingSnapshot()
    spy.mockRestore()
  })

  it('should rendering a description with the `description` prop', () => {
    render(<Banner title="test" description="test-description" />)
    expect(screen.getByText('test-description')).toBeInTheDocument()
    expect(screen.getByRole('region', {name: 'Information'})).toContainElement(screen.getByText('test-description'))
  })

  it('should support a primary action', async () => {
    const user = userEvent.setup()
    const onClick = jest.fn()
    render(
      <Banner
        title="test"
        description="test-description"
        primaryAction={<Banner.PrimaryAction onClick={onClick}>test primary action</Banner.PrimaryAction>}
      />,
    )

    expect(screen.getByRole('button', {name: 'test primary action'})).toBeInTheDocument()

    await user.click(screen.getByRole('button', {name: 'test primary action'}))
    expect(onClick).toHaveBeenCalledTimes(1)
  })

  it('should support a secondary action', async () => {
    const user = userEvent.setup()
    const onClick = jest.fn()
    render(
      <Banner
        title="test"
        description="test-description"
        secondaryAction={<Banner.SecondaryAction onClick={onClick}>test secondary action</Banner.SecondaryAction>}
      />,
    )

    expect(screen.getByRole('button', {name: 'test secondary action'})).toBeInTheDocument()

    await user.click(screen.getByRole('button', {name: 'test secondary action'}))
    expect(onClick).toHaveBeenCalledTimes(1)
  })

  it('should support primary action and secondary action', () => {
    render(
      <Banner
        title="test"
        description="test-description"
        primaryAction={<Banner.PrimaryAction>test primary action</Banner.PrimaryAction>}
        secondaryAction={<Banner.SecondaryAction>test secondary action</Banner.SecondaryAction>}
      />,
    )

    expect(screen.getByRole('button', {name: 'test primary action'})).toBeInTheDocument()
    expect(screen.getByRole('button', {name: 'test secondary action'})).toBeInTheDocument()
  })

  it('should call `onDismiss` when the dismiss button is activated', async () => {
    const user = userEvent.setup()
    const onDismiss = jest.fn()
    render(<Banner title="test" description="test-description" onDismiss={onDismiss} />)

    await user.click(screen.getByRole('button', {name: 'Dismiss banner'}))
    expect(onDismiss).toHaveBeenCalledTimes(1)

    expect(screen.getByRole('button', {name: 'Dismiss banner'})).toHaveFocus()

    await user.keyboard('{Enter}')
    expect(onDismiss).toHaveBeenCalledTimes(2)

    await user.type(screen.getByRole('button', {name: 'Dismiss banner'}), '{Space}')
    expect(onDismiss).toHaveBeenCalledTimes(3)
  })

  it('should not support onDismiss when `variant="critical"`', () => {
    const onDismiss = jest.fn()
    render(<Banner title="test" description="test-description" onDismiss={onDismiss} variant="critical" />)
    expect(screen.queryByRole('button', {name: 'Dismiss banner'})).toBe(null)
  })

  it('should pass extra props onto the container element', () => {
    const {container} = render(<Banner title="test" data-testid="test" />)
    expect(container.firstChild).toHaveAttribute('data-testid', 'test')
  })

  it('should support a custom icon for info and upsell variants', () => {
    const CustomIcon = jest.fn(() => <svg data-testid="icon" aria-hidden="true" />)
    const {rerender} = render(
      <Banner title="test" description="test-description" variant="info" icon={<CustomIcon />} />,
    )
    expect(screen.getByTestId('icon')).toBeInTheDocument()

    rerender(<Banner title="test" description="test-description" variant="upsell" icon={<CustomIcon />} />)
    expect(screen.getByTestId('icon')).toBeInTheDocument()

    rerender(<Banner title="test" description="test-description" variant="critical" icon={<CustomIcon />} />)
    expect(screen.queryByTestId('icon')).toBe(null)

    rerender(<Banner title="test" description="test-description" variant="success" icon={<CustomIcon />} />)
    expect(screen.queryByTestId('icon')).toBe(null)

    rerender(<Banner title="test" description="test-description" variant="warning" icon={<CustomIcon />} />)
    expect(screen.queryByTestId('icon')).toBe(null)
  })

  describe('Banner.Title', () => {
    it('should render as a h2 element by default', () => {
      render(
        <Banner>
          <Banner.Title>test</Banner.Title>
        </Banner>,
      )
      expect(screen.getByRole('heading', {level: 2, name: 'test'})).toBeInTheDocument()
    })

    it('should support rendering as any heading element above level 2', () => {
      const levels = [2, 3, 4, 5, 6] as const

      render(
        <>
          <Banner>
            <Banner.Title as="h2">test level 2</Banner.Title>
          </Banner>
          <Banner>
            <Banner.Title as="h3">test level 3</Banner.Title>
          </Banner>
          <Banner>
            <Banner.Title as="h4">test level 4</Banner.Title>
          </Banner>
          <Banner>
            <Banner.Title as="h5">test level 5</Banner.Title>
          </Banner>
          <Banner>
            <Banner.Title as="h6">test level 6</Banner.Title>
          </Banner>
        </>,
      )

      for (const level of levels) {
        expect(screen.getByRole('heading', {level, name: `test level ${level}`})).toBeInTheDocument()
      }
    })

    it('should support a custom `className` on the container element', () => {
      render(
        <Banner>
          <Banner.Title className="custom-class">test</Banner.Title>
        </Banner>,
      )
      expect(screen.getByRole('heading', {name: 'test'})).toHaveClass('custom-class')
    })

    it('should pass extra props onto the container element', () => {
      render(
        <Banner>
          <Banner.Title data-testid="test">test</Banner.Title>
        </Banner>,
      )
      expect(screen.getByRole('heading', {name: 'test'})).toHaveAttribute('data-testid', 'test')
    })
  })

  describe('Banner.Description', () => {
    it('should support a custom `className` on the container element', () => {
      render(
        <Banner>
          <Banner.Title>test title</Banner.Title>
          <Banner.Description className="custom-class">test description</Banner.Description>
        </Banner>,
      )
      expect(screen.getByText('test description')).toHaveClass('custom-class')
    })

    it('should pass extra props onto the container element', () => {
      render(
        <Banner>
          <Banner.Title>test title</Banner.Title>
          <Banner.Description data-testid="test">test description</Banner.Description>
        </Banner>,
      )
      expect(screen.getByText('test description')).toHaveAttribute('data-testid', 'test')
    })
  })
})
