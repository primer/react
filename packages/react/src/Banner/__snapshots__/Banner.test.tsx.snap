// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner should throw an error if no title is provided 1`] = `"Expected a title to be provided to the <Banner> component with the \`title\` prop or through \`<Banner.Title>\` but no title was found"`;
